File: atlas-sdk/python/atlas_sdk/fleet.py
from dataclasses import dataclass, field
from typing import Optional, Dict, Any, List


@dataclass
class FleetMember:
    id: str
    type: str
    status: Dict[str, Any]
    last_seen: float


@dataclass
class FleetStatus:
    members: List[FleetMember]
    healthy: int
    total: int
    mission_active: bool


@dataclass
class MissionGoal:
    description: str
    priority: int = 1


@dataclass
class MissionDefinition:
    name: str
    goals: List[MissionGoal]
    description: Optional[str] = None


class Fleet:
    """Represents a fleet of robots/drones in the Atlas system."""

    def __init__(self, client):
        self.client = client
        self.members: Dict[str, FleetMember] = {}

    def register(self, member_id: str, member_type: str) -> None:
        """Register a member in the fleet."""
        self.members[member_id] = FleetMember(
            id=member_id,
            type=member_type,
            status={"mode": "idle", "battery": 100},
            last_seen=0,
        )
        self.client.emit_event("entity:registered", "fleet", {
            "member_id": member_id,
            "type": member_type,
        })

    def unregister(self, member_id: str) -> None:
        """Remove a member from the fleet."""
        self.members.pop(member_id, None)

    async def deploy(self, mission: MissionDefinition) -> None:
        """Deploy a fleet mission."""
        await self.client.emit_remote("MISSION_RECEIVED", {
            "missionName": mission.name,
            "goalCount": len(mission.goals),
            "memberCount": len(self.members),
        })

    async def broadcast(self, signal: str, data: Optional[Dict] = None) -> None:
        """Broadcast a signal to all fleet members."""
        await self.client.emit_remote("TASK_REQUEST", {
            "name": signal,
            "data": data or {},
            "broadcast": True,
        })

    def monitor(self) -> FleetStatus:
        """Get the current fleet status."""
        healthy = sum(
            1 for m in self.members.values()
            if m.status.get("mode") != "error"
        )
        return FleetStatus(
            members=list(self.members.values()),
            healthy=healthy,
            total=len(self.members),
            mission_active=True,
        )

File: atlas-sdk/Rust/src/fleet.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use crate::client::AtlasClient;
use crate::robot::RobotStatus;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FleetMember {
    pub id: String,
    #[serde(rename = "type")]
    pub member_type: String,
    pub status: RobotStatus,
    pub last_seen: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FleetStatus {
    pub members: Vec<FleetMember>,
    pub healthy: i32,
    pub total: i32,
    pub mission_active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MissionGoal {
    pub description: String,
    pub priority: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MissionDefinition {
    pub name: String,
    pub description: Option<String>,
    pub goals: Vec<MissionGoal>,
}

pub struct Fleet {
    client: AtlasClient,
    members: HashMap<String, FleetMember>,
}

impl Fleet {
    pub fn new(client: AtlasClient) -> Self {
        Self {
            client,
            members: HashMap::new(),
        }
    }

    pub fn register(&mut self, id: &str, member_type: &str) {
        let member = FleetMember {
            id: id.to_string(),
            member_type: member_type.to_string(),
            status: RobotStatus {
                position: crate::robot::Position { x: 0.0, y: 0.0, z: 0.0 },
                battery: 100.0,
                speed: 0.0,
                mode: "idle".to_string(),
                task_count: 0,
            },
            last_seen: 0,
        };
        self.members.insert(id.to_string(), member);
    }

    pub fn unregister(&mut self, id: &str) {
        self.members.remove(id);
    }

    pub async fn deploy(&self, mission: &MissionDefinition) -> Result<(), Box<dyn std::error::Error>> {
        let mut payload = HashMap::new();
        payload.insert("missionName".to_string(), serde_json::Value::String(mission.name.clone()));
        payload.insert("goalCount".to_string(), serde_json::Value::Number(serde_json::Number::from(mission.goals.len() as i32)));
        self.client.emit_event("MISSION_RECEIVED", "Fleet", Some(serde_json::Value::Object(payload.into_iter().collect()))).await
    }

    pub async fn broadcast(&self, signal: &str, data: Option<serde_json::Value>) -> Result<(), Box<dyn std::error::Error>> {
        let mut payload = HashMap::new();
        payload.insert("name".to_string(), serde_json::Value::String(signal.to_string()));
        payload.insert("data".to_string(), data.unwrap_or(serde_json::Value::Null));
        payload.insert("broadcast".to_string(), serde_json::Value::Bool(true));
        self.client.emit_event("TASK_REQUEST", "Fleet", Some(serde_json::Value::Object(payload.into_iter().collect()))).await
    }

    pub fn monitor(&self) -> FleetStatus {
        let healthy = self.members.values().filter(|m| m.status.mode != "error").count() as i32;
        FleetStatus {
            members: self.members.values().cloned().collect(),
            healthy,
            total: self.members.len() as i32,
            mission_active: true,
        }
    }
}
